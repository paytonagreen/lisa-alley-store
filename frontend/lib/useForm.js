import { useState, useEffect } from 'react';

const useForm = (callback, initValues = {}) => {
  const [values, setValues] = useState(initValues)
  const [lowercaseValues, setLowercaseValues] = useState(initValues)
  const [isSubmitting, setIsSubmitting] = useState(false);
  
  useEffect(() => {
    if(isSubmitting) {
      callback();
    }
  }, [callback, isSubmitting]);

  const handleSubmit = e => {
    if (e) {
      e.preventDefault()
    };
    setIsSubmitting(true)
  };

  const handleChange = e => {
    e.persist();
    let val = e.target.type === 'number' ? parseFloat(e.target.value) : e.target.value;
    setValues(v => ({
      ...v, [e.target.name]: val
    }))
    setLowercaseValues(v => ({
      ...v, [e.target.name]: typeof(val) === 'string' ? val.toLowerCase() : val
    }))
    
  };

  return {
    handleChange,
    handleSubmit,
    values,
    lowercaseValues
  }
}

export default useForm;