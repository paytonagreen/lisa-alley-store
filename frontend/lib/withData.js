import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/cart/Cart'
import { LOCAL_BURGER_QUERY } from '../components/burger-menu/HamburgerMenu'

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    //local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            const data = {
              data: {cartOpen: !cartOpen},
            }
            cache.writeData(data);
            return data;
          },
          toggleBurger(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { burgerOpen } = cache.readQuery({
              query: LOCAL_BURGER_QUERY,
            });
            const data = {
              data: {burgerOpen: !burgerOpen},
            }
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
        burgerOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
