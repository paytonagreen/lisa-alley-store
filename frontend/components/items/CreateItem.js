import { useEffect, useState } from 'react';
import { useMutation, gql } from '@apollo/client';
import Router from 'next/router';

import useForm from '../../lib/useForm';
import { ALL_ITEMS_QUERY } from './Items';

import Form from '../styles/Form';
import Error from '../utils/ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
    $type: String!
    $size: String!
    $lowercaseTitle: String!
    $lowercaseDescription: String!
    $quantity: Int!
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
      type: $type
      size: $size
      lowercaseTitle: $lowercaseTitle
      lowercaseDescription: $lowercaseDescription
      quantity: $quantity
    ) {
      id
    }
  }
`;

function update(cache, payload) {
  cache.modify({
    id: cache.identify(payload.data.createItem),
    fields: {
      allItems(items, { readField }) {
        return [payload.data.createItem, ...items];
      },
    },
  });
}

function CreateItem() {
  const { values, lowercaseValues, handleChange } = useForm();
  const [image, setImage] = useState('');
  const [largeImage, setLargeImage] = useState('');
  const [lowercaseTitle, setLowercaseTitle] = useState('');
  const [lowercaseDescription, setLowercaseDescription] = useState('');

  useEffect(() => {
    setLowercaseTitle(lowercaseValues.title);
    setLowercaseDescription(lowercaseValues.description);
  }, [values.title, values.description])

  const [createItem, { loading, error }] = useMutation(CREATE_ITEM_MUTATION, {
    variables: {
      ...values,
      image,
      largeImage,
      lowercaseTitle,
      lowercaseDescription,
    },
    // update,
    refetchQueries: [{ query: ALL_ITEMS_QUERY }],
  });

  async function uploadFile(e) {
    console.log('uploading file...');
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');

    const res = await fetch(
      'https://api.cloudinary.com/v1_1/dtqqdu0so/image/upload',
      {
        method: 'POST',
        body: data,
      }
    );
    const file = await res.json();
    console.log(file);
    setImage(file.secure_url);
    setLargeImage(file.eager[0].secure_url);
  }

  return (
    <Form
      data-test="form"
      onSubmit={async (e) => {
        e.preventDefault();
        try {
          const res = await createItem();
          Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id },
          });
        } catch(error) {
          console.log(error);
        }
      }}
    >
      <Error error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="file"
            placeholder="Upload an image"
            required
            onChange={uploadFile}
          />
          {image && <img src={image} alt="Upload Preview" />}
        </label>
        <label htmlFor="title">
          Title
          <input
            type="text"
            id="title"
            name="title"
            placeholder="Title"
            required
            value={values.title}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price (In Cents)
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={values.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="type">
          Type
          <select id="type" name="type" onChange={handleChange}>
            <option value="">Select An Option</option>
            <option value="print">Print</option>
            <option value="original">Original</option>
            <option value="gallery">Gallery</option>
            <option value="shirt">Shirt</option>
            <option value="hat">Hat</option>
          </select>
        </label>
        {values.type === 'print' && (
          <label htmlFor="size">
            Size
            <select id="size" name="size" onChange={handleChange}>
              <option value="">Select An Option</option>
              <option value="11x14">11x14</option>
              <option value="30x30">30x30</option>
            </select>
          </label>
        )}
        {values.type === 'gallery' && (
          <label htmlFor="size">
            Size
            <select id="size" name="size" onChange={handleChange}>
              <option value=""></option>
              <option value="Gallery">Gallery Example</option>
            </select>
          </label>
        )}
        {values.type === 'original' && (
          <label htmlFor="size">
            Size
            <input
            type="text"
            id="size"
            name="size"
            placeholder="Size"
            required
            value={values.size}
            onChange={handleChange}
          />
          </label>
        )}
        <label htmlFor="quantity">
          Quantity
          <input
            type="number"
            id="quantity"
            name="quantity"
            placeholder="Quantity"
            required
            value={values.quantity}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          description
          <textarea
            id="description"
            name="description"
            placeholder="Enter A Description"
            required
            value={values.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
