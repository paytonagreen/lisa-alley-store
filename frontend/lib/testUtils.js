import casual from 'casual';
import React from 'react';
import { render } from '@testing-library/react';
import { ApolloProvider } from '@apollo/client';
import withData from '../lib/withData';

//setup wrapper for tests//
const Providers = ({ children, apollo }) => {
  return <ApolloProvider client={apollo} addTypename={false}>{children}</ApolloProvider>;
};

const customProviders = withData(Providers);

const customRender = (ui, options) =>
  render(ui, { wrapper: customProviders, ...options });

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
  createdAt: 1,
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  address1: casual.address1,
  address2: '',
  city: casual.city,
  state: casual.state,
  zip: 78909,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

const fakeRegularUser = () => ({
  __typename: 'User',
  id: '420',
  name: casual.name,
  email: casual.email,
  address1: casual.address,
  address2: '',

  city: casual.city,
  state: casual.state,
  zip: 11111,
  permissions: ['USER'],
  orders: [],
  cart: [],
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

const fakeCartUser = () => ({
  ...fakeUser(),
  cart: [fakeCartItem()],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeRegularUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  fakeCartUser,
  customRender as render,
};
