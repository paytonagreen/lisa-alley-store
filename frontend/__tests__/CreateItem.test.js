import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Router from 'next/router';
import { render, fakeItem } from '../lib/testUtils';
import CreateItem from '../components/items/CreateItem';

jest.mock('next/router', () => ({
  push: jest.fn(),
}));

const fillForm = async () => {
  await userEvent.type(screen.getByPlaceholderText('Title'), item.title);
  await userEvent.type(
    screen.getByPlaceholderText('Price'),
    item.price.toString()
  );
  await userEvent.type(
    screen.getByPlaceholderText('Enter A Description'),
    item.description
  );
}

const item = fakeItem();
describe('<CreateItem/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(
        <CreateItem />
    );
    expect(container).toMatchSnapshot();
  });

  it('handles state updating', async () => {
    render(
      <CreateItem />
    );

    await fillForm()

    expect(screen.getByDisplayValue(item.title)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.price.toString())).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
  });
  it('creates an item when the form is submitted', async () => {
    render(
        <CreateItem />
    );
    
    await fillForm();
    
    await userEvent.click(screen.getByText('Submit'));
    await waitFor(() => {
      expect(Router.push).toHaveBeenCalled();
      expect(Router.push).toHaveBeenCalledWith({
        pathname: '/item',
        query: { id: 'abc123' },
      });
    });
  });
});
