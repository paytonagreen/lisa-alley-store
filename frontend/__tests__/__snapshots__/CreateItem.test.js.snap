// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-phs3r-0 bOqHdm"
    data-test="form"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="file"
      >
        Image
        <input
          id="file"
          name="file"
          placeholder="Upload an image"
          required=""
          type="file"
        />
        
      </label>
      <label
        for="title"
      >
        Title
        <input
          id="title"
          name="title"
          placeholder="Title"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price (In Cents)
        <input
          id="price"
          name="price"
          placeholder="Price"
          required=""
          type="number"
          value=""
        />
      </label>
      <label
        for="type"
      >
        Type
        <select
          id="type"
          name="type"
        >
          <option
            value=""
          >
            Select An Option
          </option>
          <option
            value="print"
          >
            Print
          </option>
          <option
            value="original"
          >
            Original
          </option>
          <option
            value="gallery"
          >
            Gallery
          </option>
          <option
            value="shirt"
          >
            Shirt
          </option>
          <option
            value="hat"
          >
            Hat
          </option>
        </select>
      </label>
      <label
        for="quantity"
      >
        Quantity
        <input
          id="quantity"
          name="quantity"
          placeholder="Quantity"
          required=""
          type="number"
          value=""
        />
      </label>
      <label
        for="description"
      >
        description
        <textarea
          id="description"
          name="description"
          placeholder="Enter A Description"
          required=""
        />
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</div>
`;
